### ISSUE TITLE
Figure out a way to easily get non-package external dependencies cloned

### ISSUE THREAD

**Issue by tomvothecoder** [E3SM-Project/e3sm_to_cmip#124]
There are two non-package external dependencies that some CWL workflows rely on:

1. `metadata_path` 
  * Repository: https://github.com/E3SM-Project/CMIP6-Metadata
  * Example: https://github.com/E3SM-Project/e3sm_to_cmip/blob/413c37819dd1d7c7674f24f1ff81f61e6b7bdde3/scripts/cwl_workflows/lnd-elm/lnd-job.yaml#L22

2. `tables_path`
  * Repository: https://github.com/PCMDI/cmip6-cmor-tables
  * Example: https://github.com/E3SM-Project/e3sm_to_cmip/blob/413c37819dd1d7c7674f24f1ff81f61e6b7bdde3/scripts/cwl_workflows/lnd-elm/lnd-job.yaml#L7

Currently, users of `e3sm_to_cmip` and `esgfpub` must manually clone these repositories into their local repositories. The would then need to update the CWL workflows that they want to run to these paths.

Potential Solution(s):
1. Add a bash "getting started" script that:
  * Clones both aforementioned repositories at the same directory level as `e3sm_to_cmip`
  * Runs the `conda env create` command to create conda environments

Make sure this is included in the DevOps documentation

**Comment by tomvothecoder** [comment]
@TonyB9000 and @chengzhuzhang, I noticed that this process could be semi-automated using a bash script.

It affects `e3sm_to_cmip` and `esgfpub`/`warehouse` (https://github.com/E3SM-Project/esgfpub/issues/100).

**Comment by tomvothecoder** [comment]
Ideally, the web dev side of me would say that the `.json` files should be served over a REST API and users would query them using HTTP. However, that's another discussion to be had.

**Comment by xylar** [comment]
Would it make sense to make conda-forge data packages for these?

**Comment by xylar** [comment]
Here's an example: https://github.com/conda-forge/tzdata-feedstock

**Comment by tomvothecoder** [comment]
> Would it make sense to make conda-forge data packages for these?

If I'm guessing correctly, a conda-forge data package is essentially cloning the repository, but directly in the environment via `conda install`?

If that's the case, I'm open to the idea since it sounds like it can simplify things. I think we'd have to get the maintainers of this repo onboard: https://github.com/PCMDI/cmip6-cmor-tables

**Comment by TonyB9000** [comment]
Honestly, I would not have known that these (cmor tables path, cmip6 metadata path) were external of conda forge.  The "warehouse" (state-machine) codes are in a similar state - accessible only through git.

**Comment by xylar** [comment]
@tomvothecoder, yeah, we'd have to ask the maintainers to add a license file.  `conda-forge` wouldn't allow the recipe without it.  Other than that, we wouldn't need any permission from the developers as long as repackaging the data doesn't violate the license agreement.  I think it would be the cleanest solution in the long run but it will be some work upfront.  Happy to help.

**Comment by xylar** [comment]
@TonyB9000, you're maintaining https://github.com/E3SM-Project/CMIP6-Metadata, right?  Could you add a license file (ask @rljacob) and start making releases when things change?  If you do that, I can make a conda recipe for it on conda-forge.

**Comment by TonyB9000** [comment]
Actually, I have yet to be introduced to maintaining this repo.  Tom @tomvothecoder or Jill @chengzhuzhang  is probably the best bet here.

**Comment by xylar** [comment]
Oh, I saw that you were the most recent committer and I assumed you were the maintainer.  Oops!

**Comment by TonyB9000** [comment]
No Prob - I'm sure a future awaits me here... :)

**Comment by chengzhuzhang** [comment]
Yeah, I think re-packaging the data repo makes sense. I can take care adding a license for `CMIP6-Metadata`.
Just to be thorough, @TonyB9000 sent in email on all the external data dependencies of `esgfpub`:
cmip_metadata_path: /p/user_pub/e3sm/staging/resource/CMIP6-Metadata/
cmip_tables_path:   /p/user_pub/e3sm/staging/resource/cmip6-cmor-tables/Tables/
vrt_map_path:       /p/user_pub/e3sm/staging/resource/cmor/vrt_remap_plev19.nc
cwl_workflows_path: /p/user_pub/e3sm/staging/resource/cmor/cwl_workflows/

The bottom two are also included in `e3sm_to_cmip`.

**Comment by xylar** [comment]
I see.

So the easiest way to handle  everything here in the `scripts` directory is just to include them in the python package as well, rather than having a separate directory for them.  This would be accomplished by moving `scripts` into `e3sm_to_cmip`, adding empty `__init__.py` files in each directory, adding all the scripts to a `scripts` argument in `setup.py` (see [this example](https://github.com/MPAS-Dev/MPAS-Tools/blob/master/conda_package/setup.py#L62-L74)), and modifying the scripts so they find the `cwl_workflows` files using `importlib.resources` (see [this example](https://github.com/MPAS-Dev/compass/blob/36166e7f08c65f170e93abe11c91d8a225b802f5/compass/mpas_core.py#L58-L60)).

You will need to make sure the data files get picked up in `setup.py` as well -- by default it will only pick up python files. See [this example](https://github.com/MPAS-Dev/MPAS-Tools/blob/master/conda_package/setup.py#L61).

Given that `cwl_workflows` and `vrt_remap_plev19.nc` are not really scripts, maybe they belong in a different directory?

If someone can get the ball rolling, I can review.

**Comment by xylar** [comment]
I'm going to work on a branch with the changes I recommended above.

In that process, I noticed that the NetCDF file is already available within the package here:
https://github.com/E3SM-Project/e3sm_to_cmip/commits/master/e3sm_to_cmip/resources/vrt_remap_plev19.nc

**Comment by chengzhuzhang** [comment]
Thank you for getting start on this!
yes, this .nc files appears twice in the repo, one under /scripts, the other under /resources. 
We have been updates data under /cwl_workflows.  To package them in the proposed  data package, I'm wondering how the release schedule of the data package would be? since it requires files from three distinct repos.

**Comment by xylar** [comment]
>  To package them in the proposed data package, I'm wondering how the release schedule of the data package would be? since it requires files from three distinct repos.

These would need to be 3 distinct data packages.

**Comment by xylar** [comment]
> We have been updates data under /cwl_workflows

Could you explain what data you're referring to?  The `.cwl` files?

**Comment by xylar** [comment]
Could I get a little more clarity on the workflow involving the `.cwl` files?

Are folks going to just be running the `cwl` tool and pointing those files without modification?  Are they going to act as templates for folks to modify?  Or are folks going to be trying to modify them in place and then run with them?

The strategy for handling them would be different depending on which.

**Comment by tomvothecoder** [comment]
> Could I get a little more clarity on the workflow involving the `.cwl` files?
> 
> Are folks going to just be running the `cwl` tool and pointing those files without modification? Are they going to act as templates for folks to modify? Or are folks going to be trying to modify them in place and then run with them?
> 
> The strategy for handling them would be different depending on which.
Thanks for getting started @xylar.

From what I understand, the `.cwl` files aren't supposed to be modified by the end user. Instead, the `.yaml` file associated with the main `.cwl` file is configured. The `.yaml` files are supposed to be templates for the end user to copy and modify as needed (e.g., updating paths like the ones mentioned in this issue).

* Example `.cwl` file (static): https://github.com/E3SM-Project/e3sm_to_cmip/blob/master/scripts/cwl_workflows/lnd-elm/lnd.cwl
* Example `.yaml` file associate with the `.cwl` file (template for configuration): https://github.com/E3SM-Project/e3sm_to_cmip/blob/master/scripts/cwl_workflows/lnd-elm/lnd.cwl
* Example command: `cwltool --tmpdir-prefix=$TMPDIR ~/tomvothecoder/e3sm_to_cmip/scripts/cwl_workflows/lnd-elm/lnd.cwl ~/tomvothecoder/e3sm_to_cmip/scripts/cwl_workflows/lnd-elm/lnd-job.yaml`

Hopefully this clarifies things, but Jill can chime in too.

**Comment by xylar** [comment]
Okay, so if the `cwl` and `yaml` files were in a conda environment (under a directory like `share/e3sm_to_cmip/`, possibly available with an environment variable), the documentation could tell users how to reference the `cwl` files and copy the `yaml` files out of the conda environment and into some sort of work directory for modification?

As an example:
```
export E2C_DATA=$CONDA_PREFIX/share/e3sm_to_cmip/
cp $E2C_DATA/cwl_workflows/lnd-elm/lnd-job.yaml .
# edit lnd-job.yaml as needed
cwltool --tmpdir-prefix=$TMPDIR $E2C_DATA/cwl_workflows/lnd-elm/lnd.cwl  lnd-job.yaml
```
We could define `E2C_DATA` as part of the conda package rather than making the user do it.  It's a little bit of a pain but only a little bit.

**Comment by xylar** [comment]
Another question:  I think the scripts should be moved into the `e3sm_to_cmip` package and then made available as scripts in `setup.py`.  However, the scripts have really generic names that might be confusing to just have in the path and there would be potential for them to conflict with other packages.  Could we prefix them all with `e2c_` (or the prefix of your choice) to make it clear that they belong to `e3sm_to_cmip`?:
```
e2c_check_input_dataset.py
e2c_cmor-example.py
...
```

**Comment by xylar** [comment]
One more question: Which items in `resources` fit in each of these 3 categories?
1. we can get rid of it because it should come from an external package
2. we need to access it from `e3sm_to_cimp` the python package (or `e3sm_to_cimp` the entry point)
3. we need to access it from one of the scripts and/or the `cwl` workflows

I ask because things in category 1. need to be removed and things in category 3. aren't very accessible where they currently are.  Here is my best guess:
```
1 CMIP6_3hr.json
1 CMIP6_AERmon.json
1 CMIP6_Amon.json
1 CMIP6_CFmon.json
1 CMIP6_day.json
1 CMIP6_fx.json
1 CMIP6_Lmon.json
3? components.yaml
2 default_handler_info.yaml
2 default_metadata.json
3? e3sm_metadata.yaml
3 vrt_remap_plev19.nc
```
So it seems like only 2 of these files belong here and maybe 3 need to move somewhere outside the package so they can be made available in `share/e3sm_to_cimp/resources` in the conda environment.

**Comment by tomvothecoder** [comment]
Hey @xylar, I'm about to head out on vacation for the rest of the week. If nobody gets to them before me, I'll be happy to review your questions next week.

**Comment by xylar** [comment]
@tomvothecoder, sounds good.  Enjoy!

**Comment by xylar** [comment]
The rest of you, I have 2 branches, one of the this repo and one of the conda-forge feedstock.  They're working as expected.
https://github.com/xylar/e3sm_to_cmip/tree/move_scripts_cwl_into_package
https://github.com/xylar/e3sm_to_cmip-feedstock/tree/test_package_reorg
To try this out yourselves, you could clone both branches and then do the following in the `test_package_reorg` branch:
```
conda install mamba boa
# fix the hard-coded path to the move_scripts_cwl_into_package code
vim recipe/meta.yaml
conda mambabuild -m .ci_support/linux_64_.yaml recipe/
```
You could test this out by doing:
```
mamba install -y -n test --use-local e3sm_to_cmip=1.9.0a1
conda activate test
ls $E2C_DATA/resources
```

**Comment by xylar** [comment]
I can start mockup conda-forge recipes for the 2 data repositories while we wait on licenses and/or releases and make the test e3sm_to_cmip build depend on those as well.  That way, we could at least prove that the workflow functions as expected, fix internal paths, and begin to update documentation while we wait on those packages.

**Comment by chengzhuzhang** [comment]
Thank you Xylar! I will test. and at the same time think more about the data organization. It may not be a bad idea to fold [E3SM-Project/CMIP6-Metadata](https://github.com/E3SM-Project/CMIP6-Metadata) into $E2C_DATA. I will let you know.

**Comment by xylar** [comment]
The advantage in having https://github.com/E3SM-Project/CMIP6-Metadata be its own package is that it can have its own license and release schedule.  You could do a new release every time the data gets updated and have the version number be the date of the release.  If you tie it in to `e3sm_to_cmip`, you have to coordinate releases of the software with the changes in the metadata.

We can have a different `$E3SM_CMIP6_METADATA` variable (or similar) that points to that location, and that would get set in its own package.  I'll mock it up this evening.

**Comment by xylar** [comment]
Here are my draft conda-forge recipes for the two external data packages:
https://github.com/xylar/staged-recipes/tree/add_cmip-cmor-tables/recipes/cmip6-cmor-tables
https://github.com/xylar/staged-recipes/tree/add_e3sm-cmip6-metadata/recipes/e3sm-cmip6-metadata

You can build each of these from their respective directory within `recipes` with:
```
conda mambabuild .
```
Then, you can build the updated recipe for `e3sm_to_cmip` from https://github.com/xylar/e3sm_to_cmip-feedstock/tree/test_package_reorg, pointing to the code in https://github.com/xylar/e3sm_to_cmip/tree/move_scripts_cwl_into_package as I stated above (repeated here for convenience):
```
conda install mamba boa
# fix the hard-coded path to the move_scripts_cwl_into_package code
vim recipe/meta.yaml
conda mambabuild -m .ci_support/linux_64_.yaml recipe/
```

I made sure the internal imports are okay in the sense that the data still in `e3sm_to_cmip/resources` is getting imported from there but tables are now getting searched for using the `$CMIP6_CMOR_TABLE_ROOT` environment variable.

I looked at the YAML files in `cwl_workflows`. It seems like the best way forward there might be to treat these as Jinja templates  and translate the environment variables `$CMIP6_CMOR_TABLE_ROOT`, `$E3SM_CMIP6_METADATA` and `$E2C_DATA` into the corresponding Jinja variables (`{{ CMIP6_CMOR_TABLE_ROOT }}`, etc.) that users could then use in the YAML files.  

But I don't want to go too far down this rabbit hole.  I'm hoping you can take what I've started and (once the conda packages really exist) make it work as you like.

**Comment by chengzhuzhang** [comment]
> One more question: Which items in `resources` fit in each of these 3 categories?
> 
>     1. we can get rid of it because it should come from an external package
> 
>     2. we need to access it from `e3sm_to_cimp` the python package (or `e3sm_to_cimp` the entry point)
> 
>     3. we need to access it from one of the scripts and/or the `cwl` workflows
> 
> 
> I ask because things in category 1. need to be removed and things in category 3. aren't very accessible where they currently are. Here is my best guess:
> 
> ```
> 1 CMIP6_3hr.json
> 1 CMIP6_AERmon.json
> 1 CMIP6_Amon.json
> 1 CMIP6_CFmon.json
> 1 CMIP6_day.json
> 1 CMIP6_fx.json
> 1 CMIP6_Lmon.json
> 3? components.yaml
> 2 default_handler_info.yaml
> 2 default_metadata.json
> 3? e3sm_metadata.yaml
> 3 vrt_remap_plev19.nc
> ```
> 
> So it seems like only 2 of these files belong here and maybe 3 need to move somewhere outside the package so they can be made available in `share/e3sm_to_cimp/resources` in the conda environment.

Some of these files look familiar and they may also present in esgfpub repo. @TonyB9000 maybe you could clarify on this problem?

**Comment by xylar** [comment]
> Some of these files look familiar and they may also present in esgfpub repo. 

It would be *really* nice to avoid redundant copies of files.  That causes nightmares when they need to be updated (or keeping track of which version is being used).  I think Sterling had to do some quick-and-dirty work on this repo at times just to get the job done, but we want to clean as much of that up as we can.

**Comment by xylar** [comment]
@chengzhuzhang and @TonyB9000, could you give me some example workflows to test out my updated version of the code with?  No real rush on this but I would like to not lose momentum. I'm likely to get distracted with other work soon.

**Comment by TonyB9000** [comment]
To understand where each of these things are used (directly, or indirectly) by the warehouse state-machine, I will need to search for each, both in the filesystem (where they may be installed") and in various pieces of code (where they might be called upon.)  I can do that, it will take several hours.  Most all of these things were "integrated" by Sterling, so I must play archaeologist.

**Comment by TonyB9000** [comment]
The only item from that list that appears to be directly referenced by the warehouse is in the warehouse/resources/warehouse_config.yaml:

    vrt_map_path: /p/user_pub/e3sm/staging/resource/cmor/vrt_remap_plev19.nc

I am a bit concerned that the warehouse_config also contains these paths:

grids:
  ne30_to_180x360: /**home/zender1**/data/maps/map_ne30np4_to_cmip6_180x360_aave.20181001.nc
  oEC60to30_to_180x360: /**home/zender1**/data/maps/map_oEC60to30v3_to_cmip6_180x360_aave.20181001.nc
  ne120np4_to_cmip6_180x360: /**home/zender1**/data/maps/map_ne120np4_to_cmip6_180x360_aave.20181001.nc
  ne120np4_to_cmip6_720x1440: /**home/zender1**/data/maps/map_ne120np4_to_cmip6_720x1440_aave.20181001.nc

and under cmip_lnd_mon:
  lnd_source_grid: /**home/zender1**/data/grids/ne30np4_pentagons.091226.nc
  lnd_destination_grid: /**home/zender1**/data/grids/cmip6_180x360_scrip.20181001.nc

and under cmip_ocean_mon:
  mpas_region_path: /p/user_pub/e3sm/**baldwin32**/resources/oEC60to30v3_Atlantic_region_and_southern_transect.nc

**Comment by xylar** [comment]
@TonyB9000,

Regarding `vrt_remap_plev19.nc`, with my suggested changes, this would become:
```
  vrt_map_path: {{ E2C_DATA }}/vrt_remap_plev19.nc
```
as long as we modify the `warehouse` package to do the Jinja templating to make a few environment variables available in YAML files.

So long story short we should be able to handle that.

Regarding stuff in ` /home/zender1/`, that really is a problem.  These grids are big enough that they need to be manually placed in a known location on all E3SM supported machines (not Charlie's home directory) and we should use the `mache` package to ask which machine we're on.  Then, `mache` (or, less preferably, `warehouse`) can have machine-specific directories where the files are housed.  The file called `oEC60to30v3_Atlantic_region_and_southern_transect.nc` is outdated but there is an equivalent file for each ocean mesh in a `diagnostics` directory on each supported machine.  `mache` already knows where that directory is.

Again, here, we would need to use Jinja templating to fill in these paths in the YAML files based on which machine you're on.

This whole thing is complicated enough that we presumably need to have a tools meeting focused on it.  That could be the same meeting where we talk about `mache`, since the this issue and `mache` integration are tightly linked.
